iceoryx_hoofs/test/moduletests/test_static_lifetime_guard.cpp

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_hoofs/design_pattern/static_lifetime_guard.hpp"
+
+#include "test.hpp"
+#include <iostream>
+
+namespace
+{
+using namespace ::testing;
+
+struct Foo
+{
+    Foo()
+        : id(++instancesCreated)
+    {
+        ++ctorCalled;
+    }
+
+    ~Foo()
+    {
+        ++dtorCalled;
+    }
+
+    static void reset()
+    {
+        ctorCalled = 0;
+        dtorCalled = 0;
+    }
+
+    static uint32_t ctorCalled;
+    static uint32_t dtorCalled;
+    static uint32_t instancesCreated;
+
+    uint32_t id;
+};
+
+uint32_t Foo::ctorCalled{0};
+uint32_t Foo::dtorCalled{0};
+uint32_t Foo::instancesCreated{0};
+
+using Guard = iox::design_pattern::StaticLifetimeGuard<Foo>;
+
+// the first call to instance() creates a static instance,
+// g_instance is guarded once implicitly
+Foo& g_instance = Guard::instance();
+
+class StaticLifetimeGuard_test : public Test
+{
+  public:
+    void SetUp() override
+    {
+    }
+
+    void TearDown() override
+    {
+    }
+};
+
+// we rely on test order here
+
+TEST_F(StaticLifetimeGuard_test, staticInitializationSucceeded)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "d38b436b-f079-43fe-9d33-23d18cd08ffc");
+
+    // g_instance was constructed and the instance still exists
+    EXPECT_EQ(g_instance.id, 1);
+    EXPECT_EQ(Guard::count(), 1);
+    EXPECT_EQ(Foo::ctorCalled, 1);
+    EXPECT_EQ(Foo::dtorCalled, 0);
+}
+
+TEST_F(StaticLifetimeGuard_test, setCountWorks)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "1db790f9-d49e-44b2-b7e9-af50dd6a7d67");
+
+    auto oldCount = Guard::setCount(73);
```

Comment:
@elBoberido @FerdinandSpitzschnueffler Note that `setCount` must be used carefully as it can mess up lifetime and prevent destruction or worse undermine existing guards. Ideally it would not be possible to set it, it just exists for testing. Otherwise we cannot test statics properly.

Another option would be to derive and test a different class (but I do not like it). Maybe there is a better way, I did not think long enough about it.